enhancements/multi-arch/heterogeneous-architecture-clusters.md

```diff
@@ -0,0 +1,442 @@
+architecture. The image registry which is based on the docker distribution supports manifestlisted images, but the Imagestreams APIs around it
+do not. In order to fully support importing and local pullthrough of manifestlisted images,the imagestreams APIs would have to change to support
+manifestlists. This would include changes to:
+- `ImageStreamLayers` API which is needed to authorize pulls from the registry
+- `Image` API which would need to also contain all the child manifests in the manifestlist
+- `ImageStreamImport` API - the `ImageStreamImportStatus` would have to record the child manifests that were imported
+- `ImageStreamImage` API changes to support returning information about the child manifests.
+
+The pruner would also need some careful changes to make sure that it does not evict the manifestlisted image if any of the child manifests are in
+use. Conversely, all child manifests are also removed when an imagestream tag referencing a manifestlisted image is removed.
+
+oc tooling around imagestream commands would also need to change and factor in manifestlists in case of commands like `oc describe` which would also
+be enhanced to present the manifestlist information.
+
+This manifestlist support [enhancement][manifestlist-support] which was filed a while ago explains in detail the API changes needed and the different
+parts of the imagestreams and imageregistry functionality which would need to be changed to support manifestlisted images.
+
+### Samples
+
+As part of the changes to the Imagestreams API, the ImagestreamImport API's spec field would include a [new flag][IR-214] which would import the manifestlist
+rather than importing only an image from the manifestlist. The samples imagestreams would be updated to set this new flag. Not all samples will support
+all architectures, so in the future imagestreams would have to be annotated to specify which architectures they support. This new flag would be turned "on"
+by default importing manifestlists if the image is manifestlisted.
+
+### OLM
+
+OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.
+Today, OLM has a subscription object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would
+not be sufficient to specify filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the
+nodeselectorterms to filter on multiple values.
+
+Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
+they support through a nodeselector object which OLM will then provide the ability for the user to override if necessary. This information should
+also be added to the operands and potentially be made configurable by the user requesting the operand. The operator-sdk framework could also be
```

Comment:
so would the operands also mirror the same nodeaffinity as the operators? i assume they would have to and again this would be set by the respective teams.

> There's one very important detail in all of this: We need to ensure that if an operator says "I support x86, ARM, and Z", ALL of its related images should also support x86, ARM, and Z. Otherwise, there will be cluster configurations where the operator can run, but the operands can't.

Agreed and again this will be validated as part of the audit done by the portfolio management team

Reply:
> so would the operands also mirror the same nodeaffinity as the operators?

I think it would be up to the operator. If the operator's APIs expose a way for clients to provide nodeAffinity or if the operator has some architecture-aware logic, it could use those things to steer its workloads. 

A "dumb" operator would likely just run workloads with operand manifestlist images. That's where things get complicated though. If an operand supports x86 and arm64, and the cluster supports x86 and Z, the operator _MUST_ use nodeAffinity settings on its workloads to ensure they are scheduled on x86 nodes.

Essentially, when operator authors are declaring support for multiple architectures, that entails:
1. Making a declaration of that support via the CSV architecture annotation
2. Setting node affinity in the CSV deployment spec so that the operator is scheduled correctly.
3. Update the operator code such that when it creates/updates operand workloads, it sets those same nodeAffinity settings that are set on the operator deployment.

Reply:
i'll update the details with this